this code very faster

mProjectsCode opened the pull request:
🦀
(feel free to change the name of this PR)

mProjectsCode commented:
TODO list:

- [ ] make edge builder build the rust data directly
- [ ] relevant settings get passed to rust
- [ ] implement all features for mermaid code blocks
- [ ] implement all features for tree view
- [ ] fix all of the other views
- [ ] get incremental graph rebuilds on vault events working

SkepticMystic commented on src/components/codeblocks/CodeblockMermaid.svelte:
```diff
@@ -27,90 +27,99 @@
-	});
-
-	$: log.debug(mermaid);
+	update();
```
I think this should only be `onMount`

SkepticMystic commented on wasm/src/graph.rs:
Too big to review reasonably 😅 let's break it up and review properly before merging

mProjectsCode replied:
Yeah, this is easily doable as we can split this into multiple `impl` blocks over multiple files. But I would want to wait with it until we have a working version.

SkepticMystic reviewed (commented):
Looking good so far. I'm gonna try review frequently to keep the diff size small

SkepticMystic commented on src/commands/list_index/index.ts:
```diff
@@ -53,21 +54,23 @@ export namespace ListIndex {
 
-		tree.forEach(({ children, depth, edge }) => {
-			const display = stringify_node(edge.target_id, edge.target_attr, {
+		tree.forEach(({ children, depth, edge, node }) => {
```
I think `node` should be named more clearly. Is it the `target_node`?

SkepticMystic commented on src/components/EdgeLink.svelte:
```diff
@@ -24,9 +26,9 @@
+	cls="{cls} BC-edge {edge.implied
 		? 'BC-edge-explicit'
-		: `BC-edge-implied BC-edge-implied-${edge.attr.implied_kind}`}"
+		: `BC-edge-implied BC-edge-implied-${ 'transitive' /*edge.attr.implied_kind */}`}"
```
`implied_kind` isn't _just_ `"transitive"`. It's `"transitive:${stringify_transitive_rule(rule)}"`

SkepticMystic commented on src/components/EdgeLink.svelte:
I'd really prefer for the target_node to be on the edge itself

mProjectsCode replied:
I will implement a struct like this
```rs
pub struct EdgeStruct {
    #[wasm_bindgen(skip)]
    pub from: NodeData,
    #[wasm_bindgen(skip)]
    pub to: NodeData,
    #[wasm_bindgen(skip)]
    pub edge: EdgeData,
}
```

SkepticMystic replied:
Can you spread the `edge` data into the top-level? Something like:

```ts
type EdgeStruct = {
  source: NodeData;
  target: NodeData;
  attr: EdgeAttributes;
  // ...whhatever else is on edge
}
```
that'll align alot better with the existing type

mProjectsCode replied:
yep, i can do that

SkepticMystic commented on src/components/NestedEdgeList.svelte:
```diff
@@ -30,8 +31,8 @@
 </script>
-
-{#each tree.sort((a, b) => sort(a.edge, b.edge)) as item, i}
+<!-- TODO add sorting using the edge sorter -->
```
Why not do the sorting on JS side? Will doing it in Rust be significantly faster? Probably, huh

SkepticMystic commented on src/components/codeblocks/CodeblockTree.svelte:
Please leave a TODO here to remember to reimplement commented-out things

SkepticMystic commented on src/commands/list_index/index.ts:
```diff
@@ -54,23 +54,16 @@ export namespace ListIndex {
 			});
 
-			const link = Links.ify(node.path, display, {
+			const link = Links.ify(edge.target.path, display, {
```
Ahh I really like this :) Even more than `{ target_id, target_attr }`

SkepticMystic commented on wasm/src/graph.rs:
```diff
@@ -1212,49 +1350,55 @@ impl NoteGraph {
         if add_to_to_hash {
-            self.node_hash.insert(from_path.clone(), to_index);
+            self.node_hash.insert(source_path.clone(), target_index);
         }
```
Curious why don't you just add it to the hash where you set the flag to true, instead?

mProjectsCode replied:
Borrow checking rules. `source` and `target` point into the hashmap. If I mutate the hashmap, the references to `source` and `target` might become invalid. So I can't use `source` and `target` anymore after mutating the hashmap. So I have to make the insertions after using `source` and `target`.

SkepticMystic reviewed (commented):
Nice, continues to get closer :)

mProjectsCode commented on src/components/EdgeLink.svelte:
```diff
@@ -28,7 +28,7 @@
+	cls="{cls} BC-edge {!edge.implied
 		? 'BC-edge-explicit'
-		: `BC-edge-implied BC-edge-implied-${ 'transitive' /*edge.attr.implied_kind */}`}"
+		: `BC-edge-implied BC-edge-implied-${edge.edge_source}`}"
```
1. yes
2. `source` is already the source node

SkepticMystic replied:
Right, fair enough! I'll resolve this for now then

SkepticMystic commented on wasm/src/graph_traversal.rs:
```diff
@@ -15,24 +15,132 @@ use crate::{
+    }
+
+    pub fn equals(&self, other: &Path) -> bool {
+        self.edges == other.edges
```
Haha I'm so used to JS weirdness.. does this do a full deep equality check?

mProjectsCode replied:
yes, `==` is based on the `PartialEq` trait, which checks for deep equality (at here, technically the trait could be implemented in an arbitrary way)

SkepticMystic commented on wasm/src/graph_traversal.rs:
```diff
@@ -15,24 +15,132 @@ use crate::{
+        Path { edges }
+    }
+
+    pub fn new_start(edge: EdgeStruct) -> Path {
+        Path { edges: vec![edge] }
+    }
```
Why not just use `new`?

mProjectsCode replied:
one less function call, idk. It doesn't really make a difference.

SkepticMystic commented on src/commands/freeze_edges/index.ts:
```diff
@@ -11,13 +10,13 @@ export const freeze_implied_edges_to_note = async (
+			!e.is_self_loop() &&
+			e.implied // &&
 			// If field === null, we don't have an opposite field to freeze to
-			e.attr.field !== null,
+			// e.attr.field !== null,
```
Thanks yeah, this isn't relevant anymore

mProjectsCode commented on src/components/page_views/TrailView.svelte:
```diff
@@ -38,41 +48,49 @@
 				: plugin.settings.views.page.trail.selection === "longest"
 					? all_paths.slice(0, 1)
-					: [[]];
+					: [];
```
no, we just want an empty list of `Path`s here

mProjectsCode commented on src/components/side_views/Matrix.svelte:
```diff
@@ -24,20 +25,40 @@
-	$: sort = get_edge_sorter(edge_sort_id, plugin.graph);
+	$: sort = get_edge_sorter(edge_sort_id);
+
+	// $: edge_field_labels = resolve_field_group_labels(
+	// 	plugin.settings.edge_field_groups,
+	// 	field_group_labels,
+	// );
+
+	// $: grouped_out_edges =
+	// 	$active_file_store &&
+	// 	// Even tho we ensure the graph is built before the views are registered,
+	// 	// Existing views still try render before the graph is built.
+	// 	plugin.graph.hasNode($active_file_store.path)
+	// 		? group_by(
+	// 				plugin.graph
+	// 					.get_out_edges($active_file_store.path)
+	// 					.filter((e) =>
+	// 						has_edge_attrs(e, {
+	// 							$or_fields: edge_field_labels,
+	// 						}),
+	// 					),
+	// 				(e) => e.attr.field,
+	// 			)
+	// 		: null;
+
+	// $: sort = get_edge_sorter(edge_sort_id, plugin.graph);
```
yes, a lot of commented code can go now, but I want to make sure things work before I delete the old code

mProjectsCode commented on src/components/side_views/TreeView.svelte:
```diff
@@ -27,30 +25,43 @@
-			: [];
+	$: tree = plugin.graph.rec_traverse(
+		new TraversalOptions(
+			[$active_file_store!.path],
```
not sure, i think i copied that from the old code

mProjectsCode commented:
Some preliminary performance metrics. (Take with a grain of salt, as with all benchmarks.)
Tested with `2564` notes in a tree shape. Each note has 1 to 3 explicit `up` frontmatter links. The tree is cut off at depth 10.
Transitive rules:
```
[up] <- down
[down] <- up
[same] <- same
[uncle] <- nephew
[nephew] <- uncle
[up, down] -> same
[up, same] -> uncle
```

This fork: 
- Initial Graph: `~4ms`
- Implied Edges: `~2.5ms`
- Total: `~18ms`

Current Master (4.2.27) after running a few times to get JIT advantages:
- Implied Edges: `~70ms`
- Total: `~85ms`

This fork is about `4x` faster regarding total time and about `25x` faster in building implied edges.

mProjectsCode commented:
There might be some strangeness with sorting, as bind gen frees arrays of rust objects when passed to rust. Since there is no way to pass a reference to an array, the sorting code returns an array that has to override the array passed in.

SkepticMystic commented on wasm/src/graph_data.rs:
```diff
@@ -108,7 +108,11 @@ impl EdgeData {
 
-        result.join(" ")
+        if result.len() == 1 {
+            result[0].split('=').nth(1).unwrap().to_string()
```
Agreed

SkepticMystic commented on src/components/codeblocks/CodeblockTree.svelte:
```diff
@@ -30,19 +36,26 @@
+
 		const traversal_options = new TraversalOptions(
-			[file_path],
-			options.fields ?? [],
```
Question here. Previously the fallback was `[]`. Does your traversal treat an empty fields list as _all_ fields allowed? Cause mine needed you to explicitly pass in the full list of fields, if that's what you wanted

mProjectsCode replied:
`undefined` is all fields, `[]` is none

SkepticMystic replied:
Ok call, I'll update my change to use `undefined` when necessary

SkepticMystic commented on wasm/src/graph_data.rs:
```diff
@@ -79,21 +80,20 @@ impl EdgeData {
 
-        // TODO: maybe change the attribute options so that the JS side better matches the data
-
+        // TODO(JS): maybe change the attribute options so that the JS side better matches the data
```
I could've named it better, but with the `TODO(RUST)` comments, I more just meant TODOs left for this PR, not _where_ the todo needs to be done :)

SkepticMystic commented on wasm/src/graph_data.rs:
```diff
@@ -79,21 +80,20 @@ impl EdgeData {
             let data = match attribute.as_str() {
                 "field" => Some(("field", self.edge_type.clone())),
-                "explicit" => Some(("explicit", (!self.implied).to_string())),
+                "explicit" => Some(("explicit", (!self.explicit).to_string())),
```
Remove `!`

SkepticMystic commented on wasm/src/graph_data.rs:
```diff
@@ -79,21 +80,20 @@ impl EdgeData {
+                    if self.explicit {
                         Some(("implied_kind", self.edge_source.clone()))
                     } else {
                         None
```
Same here, add/remove `!`

SkepticMystic commented on wasm/src/graph_data.rs:
```diff
@@ -250,7 +250,7 @@ impl EdgeStruct {
     #[wasm_bindgen(getter)]
     pub fn implied(&self) -> bool {
-        self.edge.implied
+        self.edge.explicit
```
Add `!`

SkepticMystic commented on wasm/src/graph_mermaid.rs:
```diff
@@ -246,9 +245,9 @@ impl NoteGraph {
-                EitherOrBoth::Both(a, b) => a.implied && b.implied,
-                EitherOrBoth::Left(a) => a.implied,
-                EitherOrBoth::Right(b) => b.implied,
+                EitherOrBoth::Both(a, b) => a.explicit && b.explicit,
+                EitherOrBoth::Left(a) => a.explicit,
+                EitherOrBoth::Right(b) => b.explicit,
```
I don't fully understand this, but because other instances of `implied` -> `explicit` needed inversion, please double-check this

SkepticMystic commented on wasm/src/utils.rs:
```diff
@@ -1,16 +1,115 @@
+
+    pub type Logger;
+    #[wasm_bindgen(method)]
+    pub fn debug(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn info(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn warn(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn error(this: &Logger, message: &str);
```
I've also added a `feat` level for features whose purpose is to log things. It has the highest log level

SkepticMystic commented on wasm/src/utils.rs:
```diff
@@ -1,16 +1,115 @@
+    }
+
+    pub fn stop(&mut self) {
+        if self.stopped() {
+            LOGGER.warn(&format!("PerfLogger {} is already stopped", self.name));
+        } else {
```
If you double-check that it's not already stopped in this function, it feels unecessary to check this before _calling_ `self.stop()` elsewhere

mProjectsCode replied:
Kind of, the check here is to make sure that the place that is calling this checked stopped before.

SkepticMystic commented on wasm/src/utils.rs:
Nice!

SkepticMystic commented on wasm/tests/graph.rs:
```diff
@@ -60,8 +60,8 @@ fn test_implied_edge_rules_reverse_direction() {
 
-    assert_eq!(up_edge_1.weight().implied, true);
-    assert_eq!(up_edge_2.weight().implied, true);
+    assert_eq!(up_edge_1.weight().explicit, true);
+    assert_eq!(up_edge_2.weight().explicit, true);
```
Maybe I'm misunderstanding (I also don't see the full context of the file, just the diff), but shouldn't `true` now be `false`? For all instances in this file

SkepticMystic reviewed (commented):
Maybe I'm missing something, but it seems alot of the references `explicit` need to be inverted

mProjectsCode commented:
Yep, it seems i was quite tired when I made those last few commits and missed a lot of places 😅

SkepticMystic commented on src/components/codeblocks/CodeblockMermaid.svelte:
```diff
@@ -28,11 +30,19 @@
 	export const update = () => {
 		const max_depth = options.depth[1] ?? DEFAULT_MAX_DEPTH;
 
+		const source_path =  options["start-note"] || file_path || $active_file_store?.path || "";
```
`file_path` _is_ `start-note`, if present

mProjectsCode replied:
ah ok

SkepticMystic commented on src/components/side_views/MatrixEdgeField.svelte:
```diff
@@ -59,8 +58,10 @@
 							cls="font-mono"
-							label={edge.implied ? "i" : "x"}
-							aria_label={edge.get_attribute_label(show_attributes)}
+							label={edge.explicit ? "i" : "x"}
```
```suggestion
							label={edge.explicit ? "x" : "i"}
```

SkepticMystic commented on wasm/src/edge_sorting.rs:
```diff
@@ -76,16 +80,16 @@ impl EdgeSorter {
 
-        edges.sort_by(|a, b| self.apply_edge_ordering(&ordering, a, b));
+        edges.sort_by(|a, b| self.apply_edge_ordering(ordering.as_ref(), a, b));
     }
```
Are these clippy suggestions? What do they mean?

SkepticMystic commented on wasm/src/graph_mermaid.rs:
```diff
@@ -115,6 +116,16 @@ impl NoteGraph {
+            edge_sorter.sort_edges(self, &mut edge_structs);
+        }
+
+        // TODO(RUST): option to sort edges
```
```suggestion

```

SkepticMystic commented on wasm/src/graph_traversal.rs:
```diff
@@ -127,20 +106,91 @@ impl Path {
+
+#[wasm_bindgen]
+impl PathList {
+    pub fn to_paths(&self) -> Vec<Path> {
+        self.paths.clone()
+    }
```
Isn't this just a getter on `self.paths`

SkepticMystic commented on wasm/src/graph_traversal.rs:
```diff
@@ -319,6 +334,122 @@ fn rec_flatten_traversal_data(mut data: RecTraversalData, result: &mut Vec<RecTr
+
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct FlatRecTraversalResult {
```
Just at a glance, can we use this same struct to handle the `flatten` method on RecTraversalData? It seems there's some duplicated logic

mProjectsCode replied:
it's similar, but there are some differences, though the names are too close for my liking. Maybe they can be put into one method though,

SkepticMystic reviewed (commented):
Nice, I like the use of `getter_with_clone` :)

SkepticMystic reviewed (commented):
Nice, looking good

mProjectsCode commented:
Did some optimizations:
1. removed `dyn` in hot paths and replaced it with enums
2. replaced `String` with `Rc<str>` since we have lots of the same small strings, we can use `Rc`, which only clones the pointer and not the string.

mProjectsCode commented:
I limited the tree view and trail view to depth 5 for now. The graph engine can easily handle more, but the rendering is too slow. There should be an indicator that a cut-off happened though.

mProjectsCode commented:
The total graph rebuilding time is now at `~15ms` for a tree structure vault with 2.5k files.

SkepticMystic commented on src/views/matrix.ts:
```diff
@@ -22,17 +23,27 @@ export class MatrixView extends ItemView {
+
 	async onOpen() {
 		const container = this.containerEl.children[1];
 		container.empty();
```
Does `container.empty()` call `this.onClose()` to unmount the component? Or does emptying the container properly destroy the component?

mProjectsCode replied:
No, `container.empty()` only empties the element. I am not sure why the code is the way it is, seems like I didn't touch those lines.

mProjectsCode commented on src/components/side_views/Matrix.svelte:
```diff
@@ -10,28 +10,39 @@
-		// Even tho we ensure the graph is built before the views are registered,
-		// Existing views still try render before the graph is built.
-		plugin.graph.has_node($active_file_store.path)
+	let active_file = $derived($active_file_store);
```
Because stores are technically deprecated in svelte 5, and using them in certain contexts without making them a derived first can cause some strange issues like duplicated updates.

mProjectsCode commented on src/components/side_views/TreeView.svelte:
```diff
@@ -10,49 +10,79 @@
+		plugin: BreadcrumbsPlugin;
+	} = $props();
+
+	let edge_sort_id = $state(plugin.settings.views.side.tree.edge_sort_id);
+	let merge_fields = $state(plugin.settings.views.side.tree.merge_fields);
+	let show_attributes = $state(
+		plugin.settings.views.side.tree.show_attributes,
+	);
+	let show_node_options = $state(
+		plugin.settings.views.side.tree.show_node_options,
+	);
+	let field_group_labels = $state(
+		plugin.settings.views.side.tree.field_group_labels,
+	);
+	let collapse = $state(plugin.settings.views.side.tree.collapse);
```
The destructuring is wrong, as it breaks reactivity in svelte 5, but I could probably do one let view_settings = `$state(plugin.settings.views.side.tree)` and then use it like `view_settings.edge_sort_id`. That would also reduce the amount of `$state`.

mProjectsCode replied:
This should be resolved and work properly now for every component.

mProjectsCode commented:
One thing I would still like to do is look into better methods of limiting the number of items shown in views. Thanks to the nature of graphs we can get exponential blowup in regards to the depth there. This can majorly tank performance and responsiveness of the app due to too many elements being on the screen. The most promising option to me atm seems to be a hard limit 100 to 200 items that are shown.

mProjectsCode commented:
Implemented the traversal hard limit and set it to 100 for all views for now.

mProjectsCode commented:
Idk. I can look next week.
